### DISCUSSION TITLE
How do I iterate through the rolls?

### DISCUSSION THREAD

**Discussion by ialsoagree** [dice-roller/rpg-dice-roller#254] [Q&A]
I'm creating a roll using:
`roll = new DiceRoll(rollNotation);`

I then want to iterate through the individual rolls using something like:

`for (const r in roll.rolls) {
// Do stuff
}`

However, this loop only iterates once. If I assign r to a string, the string is a list of all the rolls, IE:

`const new_string = '';
new_string = '' + r; // new_string is now the string '[4, 3, 1, 5]'`

r is not an object that can be iterated over (IE. r.forEach will fail because forEach is not a function of r).

What am I doing wrong? How do I iterate over the individual rolls?

**Comment by ialsoagree** [comment]
So, I found a solution to my issue, but it may not be the ideal solution.

The first roll group is accessible via:

`roll.rolls[0].rolls.forEach(r => {
// there's r.initialValue and r.calculationValue, using standard dice I'm not sure how these differ, I assume they differ based on other engine functions })`

Accessing other roll groups gets tricky because there are modifiers in between the groups in roll.rolls. For example, 1d6+1d4 would have roll.rolls[0] contain the 1d6, roll.rolls[1] would contain the + and roll.rolls[2] would contain the 1d4.

**Reply by GreenImp** [reply]
Most of the objects returned by the dice roller have a `toString()` method, that outputs a human friendly output. ie. https://dice-roller.github.io/documentation/api/results/RollResult.html#rollresult-tostring-%E2%87%92-string

That's why you see the roll output, when you cast to a string.

The `initialValue` of a `RollResult` is rarely actually used: https://dice-roller.github.io/documentation/api/results/RollResult.html#rollresult-initialvalue-%E2%87%92-number

> The initial roll value before any modifiers.
> 
> Not used for calculations and is just for reference. You probably want `value` instead.

`calculationValue` is the value actually used in the math functions: https://dice-roller.github.io/documentation/api/results/RollResult.html#rollresult-calculationvalue-%E2%87%92-number

> The value to use in calculations. This may be changed by modifiers.

----

Based on your first example:

```javascript
roll = new DiceRoll(rollNotation);
```

The value of `roll.rolls`, it will always be an array. but the values could be several different types: https://dice-roller.github.io/documentation/api/DiceRoll.html#diceroll-rolls-%E2%87%92-array-resultgroup-rollresults-string-number

Based on your example, I'm assuming that you're rolling a single die (ie. `1d6`), so the single result you're seeing is a [`RollResults` object](https://dice-roller.github.io/documentation/api/results/RollResults.html#rollresults). You're correct that it's not iterable (Perhaps it should be). However, it has a `rolls` property, which is an array of [`RollResult` objects](https://dice-roller.github.io/documentation/api/results/RollResults.html#rollresults-rolls-%E2%87%92-array-rollresult).


Could you give me an example of the notation you're trying to roll and the end result that you're trying to achieve?
